src/ast/ast.c

```diff
@@ -527,11 +538,25 @@ char *AST_ToString(const cypher_astnode_t *node) {
 
 	char * str = (char *)cypher_astnode_get_annotation(to_string_ctx, node);
 	if(str == NULL) {
-		struct cypher_input_range range = cypher_astnode_range(node);
-		uint length = range.end.offset - range.start.offset + 2;
-		str = rm_malloc(sizeof(char) * length);
-		strncpy(str, ctx->query_data.query + range.start.offset, length - 1);
-		str[length - 1] = '\0';
+		cypher_astnode_type_t t = cypher_astnode_type(node);
+		const cypher_astnode_t *ast_identifier = NULL;
+		if(t == CYPHER_AST_NODE_PATTERN) {
+			ast_identifier = cypher_ast_node_pattern_get_identifier(node);
+		} else if(t == CYPHER_AST_REL_PATTERN) {
+			ast_identifier = cypher_ast_rel_pattern_get_identifier(node);
```

Comment:
I feel like this block could be more efficiently placed at the beginning of the function, as it does not rely on annotations at all.

Reply:
in case the identifier is null we attach anon alias and next time we return the anon alias from the annotation cox

Reply:
This is a reasonable way to avoid redundant checks on anonymous entities, I agree that this approach is fine.